packages/apps/todomvc/src/components/Root.tsx

```diff
@@ -0,0 +1,36 @@
+//
+// Copyright 2022 DXOS.org
+//
+
+import React from 'react';
+import { generatePath, Outlet, useNavigate, useParams } from 'react-router-dom';
+
+import { fromHost, fromIFrame, PublicKey } from '@dxos/client';
+import { Config, Defaults, Dynamics, Envs } from '@dxos/config';
+import { ClientProvider } from '@dxos/react-client';
+
+import { Layout } from './Layout';
+
+const configProvider = async () => new Config(await Dynamics(), await Envs(), Defaults());
+const servicesProvider = (config?: Config) =>
+  config?.get('runtime.app.env.DX_VAULT') === 'false' ? fromHost(config) : fromIFrame(config);
+
+export const Root = () => {
+  const navigate = useNavigate();
+  const { spaceKey } = useParams();
+
+  return (
+    <ClientProvider
+      config={configProvider}
+      services={servicesProvider}
+      spaceProvider={{
+        initialSpaceKey: PublicKey.safeFrom(spaceKey),
+        onSpaceChange: (spaceKey) => spaceKey && navigate(generatePath('/:spaceKey', { spaceKey: spaceKey.toHex() }))
```

Comment:
i spoke with zhenya a bit about this.
i think the API does benefit from there being a default space (perhaps stored in browser storage by the context).
but we need to be careful. the client isn't modal. we will have frames that will look at multiple spaces. we want to navigate seamlessly across spaces. so i think this is a convenience i'd prefer ideally to be factored out from "client" (even if it's bundled) and we must avoid letting this get any more "built-in" to client.
Anyway to keep and eye on and discuss.

Reply:
@zhenyasav having written the above, i'd be much happier with having another Provider inside the ClientProvider, but realize that would add some friction. perhaps though if that were combined with an "OS panels" level router?